File: sdk/python/src/agenttrunk/core/safety.py
"""AgentTrunk transport policy; copied into generated core by postprocessing."""
from contextlib import contextmanager, asynccontextmanager
import re
import httpx


def _request(client, kwargs):
    request = client.build_request(**kwargs)
    target = request.url
    if (target.scheme != 'https' and not (target.scheme == 'http' and target.host in ('localhost', '127.0.0.1', '::1'))) or target.userinfo or target.fragment:
        raise ValueError('Use a trusted HTTPS API origin (HTTP loopback only for tests)')
    if target.path.startswith('/v1/') and not re.fullmatch(r'Bearer \S+', request.headers.get('authorization', '')):
        raise ValueError('AgentTrunk access token required')
    if '/files/' in target.path and not re.fullmatch(r'[a-f0-9]{64}', target.params.get('ref', '')):
        raise ValueError('File reads require an immutable revision ID')
    return request


def _limit(request):
    return 1_000_000 if '/files/' in request.url.path else 24_000_000


def _error(response):
    request_id = response.headers.get('x-request-id', '')
    headers = {'x-request-id': request_id} if re.fullmatch(r'[A-Za-z0-9._:-]{1,200}', request_id) else {}
    return httpx.Response(response.status_code, headers=headers, content=b'', request=response.request)


def bounded_request(client, **kwargs):
    request = _request(client, kwargs)
    response = client.send(request, stream=True, follow_redirects=False)
    try:
        if not response.is_success:
            return _error(response)
        chunks = []
        size = 0
        for chunk in response.iter_bytes():
            size += len(chunk)
            if size > _limit(request):
                raise ValueError('AgentTrunk response exceeds size limit')
            chunks.append(chunk)
        # Bytes have already been decoded by httpx; remove transport encodings.
        headers = {k: v for k, v in response.headers.items() if k not in ('content-encoding', 'content-length')}
        return httpx.Response(response.status_code, headers=headers, content=b''.join(chunks), request=request)
    finally:
        response.close()


async def async_bounded_request(client, **kwargs):
    request = _request(client, kwargs)
    response = await client.send(request, stream=True, follow_redirects=False)
    try:
        if not response.is_success:
            return _error(response)
        chunks = []
        size = 0
        async for chunk in response.aiter_bytes():
            size += len(chunk)
            if size > _limit(request):
                raise ValueError('AgentTrunk response exceeds size limit')
            chunks.append(chunk)
        headers = {k: v for k, v in response.headers.items() if k not in ('content-encoding', 'content-length')}
        return httpx.Response(response.status_code, headers=headers, content=b''.join(chunks), request=request)
    finally:
        await response.aclose()


# Files are at most 1 MB, so verified reads intentionally buffer before consumption.
@contextmanager
def bounded_stream(client, **kwargs):
    response = bounded_request(client, **kwargs)
    try:
        yield response
    finally:
        response.close()


@asynccontextmanager
async def async_bounded_stream(client, **kwargs):
    response = await async_bounded_request(client, **kwargs)
    try:
        yield response
    finally:
        await response.aclose()

File: sdk/typescript/src/core/fetcher/safety.ts
/** AgentTrunk transport policy. Copied by scripts/postprocess-sdks.mjs. */
export function validateRequest(url: string, headers: Headers | Record<string, string>): void {
  const target = new URL(url);
  const loopback = ['localhost', '127.0.0.1', '[::1]'].includes(target.hostname);
  if ((target.protocol !== 'https:' && !(target.protocol === 'http:' && loopback)) || target.username || target.password || target.hash) {
    throw new Error('Use a trusted HTTPS API origin (HTTP loopback only for tests)');
  }
  const authorization = new globalThis.Headers(headers).get('authorization');
  if (target.pathname.startsWith('/v1/') && (!authorization || !/^Bearer \S+$/.test(authorization))) throw new Error('AgentTrunk access token required');
  if (target.pathname.includes('/files/') && !/^[a-f0-9]{64}$/.test(target.searchParams.get('ref') ?? '')) throw new Error('File reads require an immutable revision ID');
}

export async function boundedResponse(response: Response, url: string): Promise<Response> {
  if (!response.ok) {
    await response.body?.cancel();
    const requestId = response.headers.get('x-request-id');
    const headers = new Headers();
    if (requestId && /^[A-Za-z0-9._:-]{1,200}$/.test(requestId)) headers.set('x-request-id', requestId);
    return new Response(null, {status: response.status, headers});
  }
  if (!response.body) return response;
  const limit = new URL(url).pathname.includes('/files/') ? 1_000_000 : 24_000_000;
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let size = 0;
  try {
    while (true) {
      const {done, value} = await reader.read();
      if (done) break;
      size += value.length;
      if (size > limit) throw new Error('AgentTrunk response exceeds size limit');
      chunks.push(value);
    }
  } finally {await reader.cancel().catch(() => {}); reader.releaseLock();}
  const bytes = new Uint8Array(size);
  let offset = 0;
  for (const chunk of chunks) {bytes.set(chunk, offset); offset += chunk.length;}
  // Buffering is bounded and keeps the fetch timeout active until all bytes arrive.
  return new Response(bytes, {status: response.status, headers: response.headers});
}
